crates/wasi-http/src/body.rs

```diff
@@ -0,0 +1,513 @@
+use crate::{bindings::http::types, types::FieldMap};
+use anyhow::anyhow;
+use bytes::Bytes;
+use http_body_util::combinators::BoxBody;
+use std::{
+    convert::Infallible,
+    pin::Pin,
+    sync::{Arc, Mutex},
+    time::Duration,
+};
+use tokio::sync::{mpsc, oneshot};
+use wasmtime_wasi::preview2::{
+    self, AbortOnDropJoinHandle, HostInputStream, HostOutputStream, OutputStreamError, StreamState,
+};
+
+pub struct HostIncomingBody {
+    pub worker: AbortOnDropJoinHandle<()>,
+    pub stream: Option<HostIncomingBodyStream>,
+    pub trailers: oneshot::Receiver<Result<hyper::HeaderMap, anyhow::Error>>,
+}
+
+impl HostIncomingBody {
+    pub fn into_future_trailers(self) -> HostFutureTrailers {
+        HostFutureTrailers {
+            worker: self.worker,
+            receiver: self.trailers,
+            received: None,
+        }
+    }
+}
+
+pub struct HostIncomingBodyStream {
+    pub open: bool,
+    pub receiver: mpsc::Receiver<Result<Bytes, anyhow::Error>>,
+    pub buffer: Bytes,
+    pub error: Option<anyhow::Error>,
+}
+
+impl HostIncomingBodyStream {
+    fn new(receiver: mpsc::Receiver<Result<Bytes, anyhow::Error>>) -> Self {
+        Self {
+            open: true,
+            receiver,
+            buffer: Bytes::new(),
+            error: None,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl HostInputStream for HostIncomingBodyStream {
+    fn read(&mut self, size: usize) -> anyhow::Result<(Bytes, StreamState)> {
+        use mpsc::error::TryRecvError;
+
+        if !self.buffer.is_empty() {
+            let len = size.min(self.buffer.len());
+            let chunk = self.buffer.split_to(len);
+            return Ok((chunk, StreamState::Open));
+        }
+
+        // TODO: we need to check self.error and report it, once we have the means to do so through
+        // the streams interface.
+
+        if !self.open {
+            return Ok((Bytes::new(), StreamState::Closed));
+        }
+
+        match self.receiver.try_recv() {
+            Ok(Ok(mut bytes)) => {
+                let len = bytes.len().min(size);
+                let chunk = bytes.split_to(len);
+                if !bytes.is_empty() {
+                    self.buffer = bytes;
+                }
+
+                return Ok((chunk, StreamState::Open));
+            }
+
+            Ok(Err(e)) => {
+                self.open = false;
+                self.error = Some(e);
```

Comment:
I'm a little confused on this since stdio for example will return `Err(e)` here.  Similarly for `AsyncReadStream` it returns `Err(e)` in a case like this. 

Given the comment above though and https://github.com/bytecodealliance/wasmtime/issues/7061 I get the feeling that I'm missing context here though. This all looks correct if https://github.com/bytecodealliance/wasmtime/issues/7061 were solved, so I think I'm mostly curious at this point why stdio/`AsyncReadStream` are one way but this is a different way

Reply:
returning Err(e) in this method will trap. The stdio and AsyncReadStream places where it causes that trap by receiving an Err from the worker are bugs. We will fix those when we resolve 7061.